pkg/generators/markers.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generators
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"k8s.io/gengo/types"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+type CommentTags struct {
+	spec.SchemaProps
+
+	// Future markers can all be parsed into this centralized struct...
+	// Optional bool `json:"optional,omitempty"`
+	// Default  any  `json:"default,omitempty"`
+}
+
+func ParseCommentTags(comments []string) (CommentTags, error) {
+	// Convert the flat list of comments into a nested map by splitting keys
+	// on colons. For example, the following comments:
+	//  +foo=bar
+	//  +foo:bar=baz
+	//  +foo:bar:baz=qux
+	//  +foo:bar:baz:qux=quux
+	//
+	// Would be converted into the following map:
+	//  map[string]any{
+	//    "foo": map[string]any{
+	//      "bar": map[string]any{
+	//        "baz": map[string]any{
+	//          "qux": "quux",
+	//        },
+	//      },
+	//    },
+	//  }
+	//
+	markers, err := parseMarkers(comments, markerPrefix)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("failed to parse marker comments: %w", err)
+	}
+	nested, err := nestMarkers(markers)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("invalid marker comments: %w", err)
+	}
+
+	// Parse the map into a CommentTags type by marshalling and unmarshalling
+	// as JSON in leiu of an unstructured converter.
+	out, err := json.Marshal(nested)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("failed to marshal marker comments: %w", err)
+	}
+
+	var commentTags CommentTags
+	if err = json.Unmarshal(out, &commentTags); err != nil {
+		return CommentTags{}, fmt.Errorf("failed to unmarshal marker comments: %w", err)
+	}
+
+	return commentTags, nil
+}
+
+func parseMarkers(markerComments []string, prefix string) (map[string]any, error) {
+	markers := types.ExtractCommentTags("+", markerComments)
+
+	// Parse the values as JSON
+	result := map[string]any{}
+	for key, value := range markers {
+		if !strings.HasPrefix(key, prefix) {
+			// we only care about validation markers for now
+			continue
+		}
+
+		newKey := strings.TrimPrefix(key, prefix)
+
+		// Skip ref markers
+		if len(value) == 1 && regexp.MustCompile(`^(?:\s*ref\s{0,1}\().*\)\s*`).MatchString(value[0]) {
```

Comment:
Is this possible try `"k8s.io/gengo/examples/defaulter-gen/generators".ParseSymbolReference` against the value to see if it is a ref instead? To avoid duplciating this logic

Reply:
resolved